{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d98a7cfc",
   "metadata": {},
   "outputs": [],
   "source": [
    "![Ask Me Anything !](https://img.shields.io/badge/Ask%20me-anything-1abc9c.svg)\n",
    "[![forthebadge](https://forthebadge.com/images/badges/works-on-my-machine.svg)](https://forthebadge.com)\n",
    "\n",
    "# MINIMAP2 <br>\n",
    "---\n",
    "\n",
    "## 1. Map to a masked reference \n",
    "Minimap has worked perfectly for me to map millions of sequences into a masked referece which looks like this \n",
    "```\n",
    "> Reference_barcodes.fasta\n",
    "GCTTGGGCCGATGTCCACGAAGCTCTCCTACGNNNNNNNNNNNNNNNNNNNNNNNNNCAGTCCAGCGCCAACCAGATA\n",
    "```\n",
    "the nice think of minimap is that you do not have to create an index of the reference. \n",
    "* Map to one sample\n",
    "\n",
    "```\n",
    "minimap2 -a Reference_barcodes.fasta subset_0-1-batch_S1_L001_R1_001.fasta > alignment.sam \n",
    "```\n",
    "* Map to multiple samples using a for loop (snakemake might a better approach)\n",
    "\n",
    "\n",
    "\n",
    "References: <br>\n",
    "[1] https://github.com/lh3/minimap2/blob/master/cookbook.md#map-sr : Cookbook on minimap2 <br>\n",
    "[2] https://forthebadge.com/ : How to add a badge <br>\n",
    "[3] https://github.com/adam-p/markdown-here/wiki/Markdown-Here-Cheatsheet : Markdown Cheatseat <br>\n",
    "\n",
    "# SAMTOOLS <br>\n",
    "---\n",
    "\n",
    "## Installing samtools \n",
    "\n",
    "```r\n",
    "cd ~\n",
    "# optional. you may already have a src directory\n",
    "mkdir src\n",
    "cd ~/src\n",
    "git clone https://github.com/samtools/htslib\n",
    "git clone https://github.com/samtools/samtools\n",
    "cd samtools\n",
    "make\n",
    "cp samtools ~/bin\n",
    "```\n",
    "\n",
    "now put in your folder the a sam file. My sam file is_ **barcodes_alignment.sam**_\n",
    "\n",
    "\n",
    "## 1. Check the header of your sam file \n",
    "\n",
    "```\n",
    "head barcodes_alignment.sam\n",
    "\n",
    ">\n",
    "@SQ\tSN:Reference_barcodes\tLN:172\n",
    "@PG\tID:minimap2\tPN:minimap2\tVN:2.15-r905\tCL:minimap2 -a Reference_barcodes.fasta subset_0-1-batch_S1_L00\n",
    "```\n",
    "your header should start with the sign \"@\",  which is an indicator for a header line. If you don't see lines starting with the \"@\" sign, the header information is most likely missing.\n",
    "\n",
    "If the header is absent from the SAM file use the command below, where reference.fa is the reference fasta file used to map the reads:\n",
    "\n",
    "samtools view -bT Reference_barcodes.fasta barcodes_alignment.sam > barcodes_alignment.bam\n",
    "\n",
    "## 2. Convert your sam file to bam file \n",
    "\n",
    "```\n",
    "samtools view -S -b barcodes_alignment.sam > barcodes_alignment.bam\n",
    "```\n",
    "S indicates that give it a sam file \n",
    "b indicates that you want bam out of it\n",
    "\n",
    "## 3. Sort the bam file \n",
    "\n",
    "The sequences have the same order as in the fasta file that you used to map the reference. Always sort your BAM files; many downstream programs only take sorted BAM files.\n",
    "\n",
    "```\n",
    "samtools sort barcodes_alignment.bam -o barcodes_alignment_sorted.bam\n",
    "```\n",
    "## 4. Index the sorted bam file \n",
    "A BAM index file is usually needed when visualising a BAM file.\n",
    "\n",
    "```\n",
    "samtools index barcodes_alignment_sorted.bam > index_barcodes_alignment_sorted.bam\n",
    "```\n",
    "\n",
    "**Important** : Use biosyntaxt to beautify the outcome of your sam file https://github.com/bioSyntax/bioSyntax\n",
    "samtools view -h NA19238.bam | sam-less -\n",
    "\n",
    "## 5. Count the mapped reads \n",
    "\n",
    "* count **all reads** in the bam file \n",
    "\n",
    "```\n",
    "samtools idxstats index_barcodes_alignment_sorted.bam | awk '{s+=$3+$4} END {print s}'\n",
    "```\n",
    "\n",
    "* count only the **mapped** reads in the bam file\n",
    "```\n",
    "samtools idxstats index_barcodes_alignment_sorted.bam | awk '{s+=$3} END {print s}'\n",
    "\n",
    "samtools view -c -F 4 index_barcodes_alignment_sorted.bam\n",
    "```\n",
    "\n",
    "* count only the **unmapped** reads in the bam file \n",
    "\n",
    "samtools view -c -f 4 index_barcodes_alignment_sorted.bam\n",
    "\n",
    "## 6. Plot a depth of coverage across a bam file \n",
    "\n",
    "[1] https://labs.epi2me.io/notebooks/Introduction_to_SAM_and_BAM_files.html\n",
    "\n",
    "## 7. Remove sequences that map to a specific area in the genome \n",
    "\n",
    "bedtools intersect -abam file.bam -b filter.bed -v > filtered.bam\n",
    "filter.bed should contain\n",
    "\n",
    "chr    start     end\n",
    "\n",
    "and maybe \n",
    "samtools view input.bam -b -h -o output_inRegions.bam -U output_outRegions.bam -L Regions.bed\n",
    "\n",
    "\n",
    "\n",
    "Other important references: <br>\n",
    "[1] http://quinlanlab.org/tutorials/samtools/samtools.html  : Notes from Quinlan lab <br>\n",
    "[2] https://davetang.org/wiki/tiki-index.php?page=SAMTools  : Notes from wiki cms <br>\n",
    "[3] http://www.novocraft.com/documentation/novoalign-2/novoalign-ngs-quick-start-tutorial/1040-2/ : extract unmapped reads <br>\n",
    "[4] https://www.metagenomics.wiki/tools/samtools/number-of-reads-in-bam-file : count unmapped reads, metagenomics notes <br>\n",
    "[5] http://www.bioinf.uni-leipzig.de/publications/supplements/13-008 : reuse unmapped reads with segemehl <br>\n",
    "[6] https://davetang.github.io/learning_bam_file/#filtering-unmapped-reads : Notes from dave tang <br>\n",
    "[7] https://qnot.org/2012/04/14/counting-the-number-of-reads-in-a-bam-file/ : Count number of reads thats very interesting <br>\n",
    "[8] https://comppopgenworkshop2019.readthedocs.io/en/latest/contents/02_bam_files/bam_files.html : Impressive Converting, Filtering SAM and BAM Files\n",
    "\n",
    "\n",
    "# Comments: \n",
    "\n",
    "## Use taxonomic profile softwares in your raw data \n",
    "You can use \n",
    "* [Kraken2](https://github.com/DerrickWood/kraken2/wiki)\n",
    "* [Kaiju](https://github.com/bioinformatics-centre/kaiju) <br>\n",
    "and you can visualise everything using <br>\n",
    "* [Kronatools](https://github.com/marbl/Krona)\n",
    "\n",
    "### Kraken2 - Krona tools\n",
    "1. Make a database \n",
    "```\n",
    "kraken2-build --standard --threads 24 --db kraken_db\n",
    "```\n",
    "2. run your data over the db\n",
    "```\n",
    "kraken2 --paired --threads {threads} -db {params.db} --confidence 0.5 --output output/ cleandata/sample1_clean_R1.fq cleandata/sample1_clean_R2.fq ... cleandata/samplen_clean_R1.fq cleandata/samplen_clean_R2.fq\n",
    "```\n",
    "3. use kraken output to Krona \n",
    "```\n",
    "ktImportTaxonomy -q 2 -t 3 Sample1.txt Sample2.txt -o krona.html \n",
    "```\n",
    "References: <br>\n",
    "[1] https://genomics.sschmeier.com/ngs-taxonomic-investigation/index.html : thats the best kraken turorial I ve found so far\n",
    "\n",
    "### Merge files from different lanes \n",
    "If your files come into multiple lanes you can concatenate them using the following code: \n",
    "\n",
    "```\n",
    "for i in $(find ./ -type f -name \"*.fastq.gz\" | while read F; do basename $F | rev | cut -c 22- | rev; done | sort | uniq)\n",
    "\n",
    "    do echo \"Merging R1\"\n",
    "\n",
    "cat \"$i\"_L00*_R1_001.fastq.gz > \"$i\"_ME_L001_R1_001.fastq.gz\n",
    "\n",
    "       echo \"Merging R2\"\n",
    "\n",
    "cat \"$i\"_L00*_R2_001.fastq.gz > \"$i\"_ME_L001_R2_001.fastq.gz\n",
    "\n",
    "done\n",
    "```\n",
    "\n",
    "Also check the following commands which might be very useful to you <br>\n",
    "1. https://unix.stackexchange.com/questions/394479/concatenating-multiple-fastq-files <br>\n",
    "2. https://unix.stackexchange.com/questions/436771/how-to-concatenate-rna-seq-files-generated-in-differnt-lanes <br>\n",
    "3. https://unix.stackexchange.com/questions/615815/for-loop-to-catenate-files-with-two-variables <br>\n",
    "\n",
    "\n",
    "### Copy-paste the contents of folder to another folder \n",
    "\n",
    "```\n",
    "cp -a /source/. /dest/\n",
    "```\n",
    "\n",
    "The -a option is an improved recursive option, that preserve all file attributes, and also preserve symlinks.\n",
    "\n",
    "The . at end of the source path is a specific cp syntax that allow to copy all files and folders, included hidden ones.\n",
    "\n",
    "\n",
    "\n",
    "## Add your favourite websites that contain bioinfromatic software \n",
    "\n",
    "https://astrobiomike.github.io/unix/installing_tools#my-bioinformatics-tools-bit <br>\n",
    "https://astrobiomike.github.io/unix/modifying_your_path <br>\n",
    "http://maasha.github.io/biopieces/ <br>\n",
    "\n",
    "\n",
    "## How to find tRNAs and rRNAs in your genome? \n",
    "\n",
    "+ To find tRNAs you can use <br>\n",
    "    - A BLAT search in the USSC browser https://genome.ucsc.edu/cgi-bin/hgBlat <br>\n",
    "    - tRNA scan is useful for sequences less than a 1Million bp http://trna.ucsc.edu/tRNAscan-SE/ <br>\n",
    "    - tRNA finder seems crap https://ei4web.yz.yamagata-u.ac.jp/~kinouchi/tRNAfinder/\n",
    "\n",
    "+ To find rRNAs you can use <br>\n",
    "    - barrnap https://github.com/tseemann/barrnap\n",
    "    - https://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-021-04316-z this paper contains a lot of info about rRNA prediction\n",
    "    - BLAST using an rRNA database\n",
    "\n",
    "\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
